File: app/src/main/java/com/gameautopilot/app/brain/Brain.kt
package com.gameautopilot.app.brain

import com.gameautopilot.app.core.Action
import com.gameautopilot.app.core.MarkBox

interface Brain {
    suspend fun decide(ctx: BrainContext): BrainDecision
}

data class BrainContext(
    val gameName: String,
    val gamePackage: String,
    val gameSystemPrompt: String,
    val screenWidth: Int,
    val screenHeight: Int,
    val screenshotBase64Jpeg: String,
    val ocrLines: List<String>,
    val a11yLines: List<String>,
    val marks: List<MarkBox>,
    val recentActionLabels: List<String>,
    val stuckHint: String? = null,
    /** Persistent per-game notes carried across ticks (and app restarts) — see GameMemoryStore. */
    val gameMemory: String = ""
)

data class BrainDecision(
    val thought: String,
    val actions: List<Action>,
    val confidence: Double,
    /** Non-null when the brain wants to replace gameMemory for future ticks. */
    val memoryUpdate: String? = null
) {
    companion object {
        val EMPTY = BrainDecision("", emptyList(), 0.0)
    }
}

class BrainException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)

File: mac/Sources/GameAutopilotMac/Brain/Brain.swift
import Foundation

protocol Brain {
    func decide(_ ctx: BrainContext) async throws -> BrainDecision
}

struct BrainContext {
    let targetName: String
    let targetBundleIdentifier: String
    let targetPrompt: String
    let goalMode: GoalMode
    let screenWidth: Int
    let screenHeight: Int
    let screenshotBase64Jpeg: String
    let ocrLines: [String]
    let a11yLines: [String]
    let marks: [MarkBox]
    let recentActionLabels: [String]
    let stuckHint: String?
    /// Persistent per-target notes carried across ticks (and app restarts) -- see TargetMemoryStore.
    let memory: String
    /// False when a different app is currently frontmost (an ad redirect,
    /// accidental navigation, or the user deliberately switched away).
    /// See DecisionLoop's bounded-recovery handling.
    let targetIsFrontmost: Bool
}

struct BrainDecision {
    let thought: String
    let actions: [Action]
    let confidence: Double
    /// Non-nil when the brain wants to replace `memory` for future ticks.
    let memoryUpdate: String?
    /// True when the brain believes the target's goal has been reached.
    /// Only meaningful when the target's goalMode is .playUntilComplete --
    /// PromptBuilder instructs the brain to never set this for .keepRunning
    /// targets, which have no completion state by definition.
    let goalComplete: Bool

    static let empty = BrainDecision(thought: "", actions: [], confidence: 0, memoryUpdate: nil, goalComplete: false)
}

/// LocalizedError, not just CustomStringConvertible: conforming to
/// CustomStringConvertible alone does NOT make `.localizedDescription`
/// use `message` -- that's a different mechanism, and the default
/// Error.localizedDescription would otherwise print a generic
/// "operation couldn't be completed" string instead.
struct BrainError: Error, CustomStringConvertible, LocalizedError {
    let message: String
    var description: String { message }
    var errorDescription: String? { message }
    init(_ message: String) { self.message = message }
}
